R/zchunk_LA1231.elec_tech.R

```diff
@@ -37,59 +38,168 @@ module_energy_LA1231.elec_tech_DISABLED <- function(command, ...) {
+    # Therefore there is not efficiency values for every historical year. Efficiencies for all historical years are needed
+    # To estimate outputs from fuel/technology by region in the electricity sector
+    tibble(supplysector = "electricity", subsector="gas", year = HISTORICAL_YEARS) %>%
+      mutate(year = as.numeric(year)) %>%
+      full_join(energy.GAS_TECH, by = "supplysector") -> Aux_gas_tech_elec
+
+    # Perform interpolation for gas technologies efficiencies , max, and rate
+    # Interpolation is needed since efficiencies are given for 1971, 1990, 2005 and 2010 only. Not for every historical year.
+    A23.globaltech_eff %>%
+      filter(subsector == "gas") %>%
+      semi_join(calibrated_techs, by = c("supplysector", "subsector", "technology")) %>%
+      gather(year, efficiency_tech, -supplysector, -subsector, -technology, -minicam.energy.input, -improvement.max, -improvement.rate, -improvement.shadow.technology) %>%
+      mutate(year = as.numeric(year)) %>%
+      mutate(efficiency_tech = as.numeric(efficiency_tech)) %>%
+      semi_join(Aux_gas_tech_elec, by = "year") %>%
+      full_join(Aux_gas_tech_elec, by = c("supplysector", "subsector", "technology","minicam.energy.input","year")) %>%
+      group_by(supplysector, subsector, technology) %>%
+      mutate(efficiency_tech = approx_fun(year, efficiency_tech)) %>%
+      mutate(improvement.max = approx_fun(year, improvement.max)) %>%
+      mutate(improvement.rate = approx_fun(year, improvement.rate)) -> L1231.eff_R_elec_gas_tech
+
+    # Reset upper and lower bound efficiencies, as needed
+    # Where avg efficiency is outside of the range of the two technologies, set the lower bound equal to the average, and the upper equal to
+    # the average plus a small adjustment. This adjustment avoids assigning 100% of market share to combined-cycle.
+    # NOTE: this is complicated. Heat output is not allowed for combined cycle power plants--it is assumed that CHP plants are all single-cycle.
+    # However if a region has very high efficiencies of CC power plants that bring the weighted average higher than our assumed CC efficiency (e.g. Turkey),
+    # this could set the market share of single-cycle power plants to 0 and therefore zero out any heat production from these technologies.
+    # Adding this small adjustment factor to the efficiencies avoids this outcome:
+    # Small adjustment factor for elec efficienies (explaining its need in the above comment)
+    ELEC_ADJ <- 0.03
+
+    L1231.eff_R_elec_gas_Yh_gathered %>%
+      full_join(select(filter(L1231.eff_R_elec_gas_tech, technology == "gas (steam/CT)"), year,efficiency_tech, supplysector, subsector, technology),by="year") %>%
+      select(-supplysector, -subsector, -technology) %>%
+      left_join(select(filter(L1231.eff_R_elec_gas_tech, technology == "gas (CC)"), year, efficiency_tech, supplysector, subsector, technology), by = "year") %>%
+      select(-supplysector, -subsector, -technology) %>%
```

Comment:
Ok, Sorry I missed the rest of the hard coded technologies. As you suggested, I created the constant GAS_TECH1 and GAS_TECH2 from the unique values of gas technologies in calibrated_techs.csv. This works now. The only issue I see is that if we want to add a new technology in the future, then this chunk has to be modified (since this approach just considers 2 technologies).